runtime/complex.go

```diff
@@ -68,5 +70,190 @@ func complexRepr(f *Frame, o *Object) (*Object, *BaseException) {
+// on whether the source value was too large or too small. Note that if the
+// source number is an infinite float, the result will be infinite without
+// overflow, (+-Inf, true).
+// If the input is not a number, it will return (0, false).
+func complexCoerce(o *Object) (complex128, bool) {
+	if o.isInstance(ComplexType) {
+		return toComplexUnsafe(o).Value(), true
+	}
+	floatO, ok := floatCoerce(o)
+	if !ok {
+		if math.IsInf(floatO, 0) {
+			return complex(floatO, 0.0), false
+		}
+		return 0, false
+	}
+	return complex(floatO, 0.0), true
+}
+
+func complexNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc == 0 {
+		return newObject(t), nil
+	}
+	if argc > 2 {
+		return nil, f.RaiseType(TypeErrorType, "'__new__' of 'complex' requires at most 2 arguments")
+	}
+	if t != ComplexType {
+		// Allocate a plain complex then copy it's value into an object
+		// of the complex subtype.
+		x, raised := complexNew(f, ComplexType, args, nil)
+		if raised != nil {
+			return nil, raised
+		}
+		result := toComplexUnsafe(newObject(t))
+		result.value = toComplexUnsafe(x).Value()
+		return result.ToObject(), nil
+	}
+	o := args[0]
+	if argc == 1 {
+		if complexSlot := o.typ.slots.Complex; complexSlot != nil {
+			result, raised := complexSlot.Fn(f, o)
+			if raised != nil {
+				return nil, raised
+			}
+			if !result.isInstance(ComplexType) {
+				format := "__complex__ returned non-complex (type %s)"
+				return nil, f.RaiseType(TypeErrorType, fmt.Sprintf(format, result.typ.Name()))
+			}
+			return result, nil
+		}
+		if floatSlot := o.typ.slots.Float; floatSlot != nil {
+			result, raised := floatSlot.Fn(f, o)
+			if raised != nil {
+				return nil, raised
+			}
+			if !result.isInstance(FloatType) {
+				exc := fmt.Sprintf("__float__ returned non-float (type %s)", result.typ.Name())
+				return nil, f.RaiseType(TypeErrorType, exc)
+			}
+			f := toFloatUnsafe(result).Value()
+			return NewComplex(complex(f, 0)).ToObject(), nil
+		}
+		if !o.isInstance(StrType) {
+			return nil, f.RaiseType(TypeErrorType, "complex() argument must be a string or a number")
+		}
+		s := toStrUnsafe(o).Value()
+		result, err := parseComplex(s)
+		if err != nil {
+			return nil, f.RaiseType(ValueErrorType, fmt.Sprintf("could not convert string to complex: %s", s))
+		}
+		return NewComplex(result).ToObject(), nil
+	}
+
+	// TODO: fix me
+	return nil, nil
+}
+
+// ParseComplex converts the string s to a complex number.
+// If string is well-formed (one of these forms: <float>, <float>j,
+// <float><signed-float>j, <float><sign>j, <sign>j or j, where <float> is
+// any numeric string that's acceptable by strconv.ParseFloat(s, 64)),
+// ParseComplex returns the respective complex128 number.
+func parseComplex(s string) (complex128, error) {
+	ts := strings.Trim(s, "() ")
+	if !strings.Contains(ts, "j") {
+		result, err := strconv.ParseFloat(ts, 64)
+		return complex(result, 0), err
+	}
+	ts = strings.Replace(ts, "j", "", -1)
```

Comment:
It's possible this would be cleaner by parsing the expression as a regular expression. According to the [CPython source](https://github.com/python/cpython/blob/2.7/Objects/complexobject.c#L1021), the pattern is basically `<float>|<float>j|<float><signed float>j|<float>[+-]j|[+-]j|j` where floats have the form given here: https://docs.python.org/2/reference/lexical_analysis.html#floating-point-literals

It's not exactly simple, but I think the code would be easier to reason about. As it stands, it's pretty difficult to understand what's going on here.